dashboard/app/controllers/script_levels_controller.rb

```diff
@@ -118,22 +118,36 @@ def hidden_stage_ids
 
   # toggles whether or not a stage is hidden for a section
   def toggle_hidden
+    script_id = params.require(:script_id)
     section_id = params.require(:section_id).to_i
-    stage_id = params.require(:stage_id)
+    stage_id = params[:stage_id]
     # this is "true" in tests but true in non-test requests
     should_hide = params.require(:hidden) == "true" || params.require(:hidden) == true
 
     section = Section.find(section_id)
     authorize! :read, section
 
-    # TODO(asher): change this to use a cache
-    return head :forbidden unless Stage.find(stage_id).try(:script).try(:hideable_stages)
+    if stage_id
+      # TODO(asher): change this to use a cache
+      return head :forbidden unless Stage.find(stage_id).try(:script).try(:hideable_stages)
 
-    hidden_stage = SectionHiddenStage.find_by(stage_id: stage_id, section_id: section_id)
-    if hidden_stage && !should_hide
-      hidden_stage.delete
-    elsif hidden_stage.nil? && should_hide
-      SectionHiddenStage.create(stage_id: stage_id, section_id: section_id)
+      hidden_stage = SectionHiddenStage.find_by(stage_id: stage_id, section_id: section_id)
+      if hidden_stage && !should_hide
+        hidden_stage.delete
+      elsif hidden_stage.nil? && should_hide
+        SectionHiddenStage.create(stage_id: stage_id, section_id: section_id)
+      end
+    else
+      # We don't have a stage id, implying we instead want to toggle the hidden state of this script
+      script = Script.get_from_cache(script_id)
+      return head :forbidden if script.nil?
+
+      hidden_script = SectionHiddenScript.find_by(script_id: script.id, section_id: section_id)
```

Comment:
We (at least @joshlory and I) have been adopting a 'fat model, thin controller' philosophy when writing Rails code.  I wonder if this would all be cleaner if we added `toggle_hidden` methods to the `Stage` and `Script` models, and let them own validation and consistency checks.  Maybe ask someone better at Rails than me?